include/netuit/topology/Topology.hpp

```diff
@@ -2,11 +2,312 @@
+          std::begin(outputs),
+          std::end(outputs)
+        );
+      }
+    );  return std::make_pair(x_adj, adjacency);
+  }
+
+  /// Apply METIS' K-way partitioning algorithm to subdivide topology
+  /// @param parts number of parts to subdivide topology into
+  /// @return vector indicating what partition each vertex should go into
+  emp::vector<idx_t> Optimize(int32_t parts) const {
+    // set up variables
+    idx_t nodes = topology.size();
+    idx_t n_cons = 1;
+    idx_t volume;
+
+    // get topology as CSR
+    auto [xadj, adjacency] = AsCSR();
+
+    // set up result vector
+    emp::vector<idx_t> result(nodes);
+
+    // set up options array
+    // in this case, it's just the default options
+    idx_t options[METIS_NOPTIONS];
+    METIS_SetDefaultOptions(options);
+
+    // call partitioning algorithm
+    int status = METIS_PartGraphKway(
+      &nodes, // number of vertices in the graph
+      &n_cons, // number of balancing constraints.
+      xadj.data(), // array of node indexes into adjacency[]
+      adjacency.data(), // array of adjacenct nodes for every node
+      nullptr, // weights of nodes
+      nullptr, // size of nodes for total comunication value
+      nullptr, // weights of edges
+      &parts, // number of parts to partition the graph into
+      nullptr, // weight for each partition and constraint
+      nullptr, // allowed load imbalance tolerance for each constraint
+      nullptr, // array of options
+      &volume, // edge-cut or total comm volume of the solution
+      result.data() // partition vector of the graph
+    );
+    // deal with return code
+    uitsl::metis::verify(status);
+
+    return result;
+  }
+
+  void PrintEdgeList(std::ostream& os = std::cout) const noexcept {
+    // FIX THIS
+    // node1 node2
+    // node1 node3
+    // etc
+    for (size_t i = 0; i < topology.size(); ++i) {
+      for (const auto& output : topology[i].GetOutputs()) {
+        os << i << " " << output << std::endl;
+      }
+    }
+  }
+
+  void PrintAdjacencyList(std::ostream& os = std::cout) const noexcept {
+    for (size_t i = 0; i < topology.size(); ++i) {
+      os << i;
+      for (const auto& output : GetNodeOutputs(topology[i])) {
+        os << " " << output;
+      }
+      os << std::endl;
+    }
+  }
+
+  void PrintPartition(const emp::vector<int>& partition, std::ostream& os = std::cout) const {
+    auto name_node = [&partition] (const size_t index) -> std::string {
+      return emp::keyname::pack({
+        {"node_id", emp::to_string(index)},
+        {"partition", emp::to_string(partition[index])}
+      });
+    };
+
+    for (size_t i = 0; i < topology.size(); ++i) {
+      os << name_node(i) << " ";
+      os << topology[i].GetNumOutputs() << std::endl;
+
+      for (const auto& node : GetNodeOutputs(topology[i])) {
+        os << name_node(node) << std::endl;
+      }
+    }
+  }
+
+  Topology GetSubTopology(const std::unordered_set<size_t>& node_ids) const {
```

Comment:
Can you add a docstring or comment explaining what this does / when we want to use it?